File: vet_care/vet_care/report/patient_history/patient_history.py
# Copyright (c) 2013, 9T9IT and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
from frappe import _
import frappe


def execute(filters=None):
	columns = _get_columns(filters)
	data = _get_data(filters)
	return columns, data


def _get_columns(filters):
	def make_column(label, fieldname, width, fieldtype='Data', options=''):
		return {
			'label': _(label),
			'fieldname': fieldname,
			'fieldtype': fieldtype,
			'width': width,
			'options': options
		}
	return [
		make_column('Posting Date', 'posting_date', 130, 'Date'),
		make_column('Activity', 'activity_type', 80, 'Data'),
		make_column('Description', 'description', 893, 'Data')
	]


def _get_data(filters):
	return _get_patient_activity_items(filters.get('patient'))


def _get_patient_activity_items(patient):
	return frappe.db.sql("""
		SELECT
			doc.posting_date,
			item.activity_type,
			item.description
		FROM `tabPatient Activity Item` item
		INNER JOIN `tabPatient Activity` doc ON item.parent = doc.name
		WHERE doc.patient = %s
	""", patient, as_dict=True)

File: vet_care/vet_care/report/patient_history/patient_history.js
// Copyright (c) 2016, 9T9IT and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Patient History"] = {
	"filters": [
        {
            fieldname: 'patient',
            label: __('Patient'),
            fieldtype: 'Link',
            options: 'Patient',
            on_change: function(report) {
                _set_patient_information(report);
                report.refresh();
            }
        },

        // metadata for html print
        {
            fieldname: 'patient_name',
            fieldtype: 'Data',
            hidden: true,
        },
        {
            fieldname: 'customer_name',
            fieldtype: 'Data',
            hidden: true,
        },
        {
            fieldname: 'patient_dob',
            fieldtype: 'Date',
            hidden: true,
        },
        {
            fieldname: 'species',
            fieldtype: 'Data',
            hidden: true,
        },
        {
            fieldname: 'breed',
            fieldtype: 'Data',
            hidden: true,
        },
        {
            fieldname: 'color',
            fieldtype: 'Data',
            hidden: true,
        },
        {
            fieldname: 'chip_id',
            fieldtype: 'Data',
            hidden: true,
        },
        {
            fieldname: 'neutered',
            fieldtype: 'Data',
            hidden: true,
        },
        {
            fieldname: 'deceased',
            fieldtype: 'Data',
            hidden: true
        },
        {
            fieldname: 'posting_date',
            fieldtype: 'Date',
            hidden: true,
            default: frappe.datetime.get_today(),
        }
	]
};

async function _set_patient_information(report) {
    const patient_information = await frappe.db.get_doc('Patient', report.get_filter_value('patient'));
    const customer_name = await frappe.db.get_value('Customer', patient_information.customer, 'customer_name');
    const mapped_fields = {
        'patient_name': patient_information.patient_name,
        'customer_name': customer_name,
        'patient_dob': patient_information.dob,
        'color': patient_information.vc_color,
        'species': patient_information.vc_species,
        'breed': patient_information.vc_breed,
        'chip_id': patient_information.vc_chip_id,
        'neutered': patient_information.vc_neutered,
        'deceased': patient_information.vc_deceased ? 'Yes' : 'No'
    };
    Object.keys(mapped_fields).forEach((key) =>
        report.set_filter_value(key, mapped_fields[key])
    );
}
